feat(api): gripper pickup from lifted thermocycler position

fsinapi opened the pull request:
# Overview

This PR changes the gripper behavior to pick up plates from a "lifted" position from the Thermocycler Gen2. This PR also _removes_ the "gripper wiggle" functionality for dislodging plates, because the lifted position provides a much better experience by increasing the reliability of TC2 plate transfers.

The `thermocycler_plate_lifter` is changed to a context manager to be able to safely open _and_ close the thermocycler lid during a gripper movement. Specifically, the behavior is the following:
* The TC2 performs the plate lift movement (which is a command that will push the plate up off the surface and then drop it back down)
* The TC2 moves the lid back _up_ to the maximum point, presenting the plate for the gripper
* The gripper picks up the plate from the elevated position
* Once the plate has been moved to a new location, the thermocycler lid is returned to the normal "open" position

An important note: THE PICKUP OFFSET FOR THE TC2 NOW GETS AN AUTOMATIC `z=3.5` INCREASE. This is baked into the protocol engine right now, it must NOT be added in the protocols.

# Test Plan

Ran a script that moves a plate from a TC2 to another deck slot with the gripper. The TC2 now opens the lid to present the plate, and moves the lid back down after the plate is moved.

# Changelog

- Added TC driver code to "jog" the lid using a debugging gcode (already supported in all released TC2 firmware)
- Changed thermocycler plate lifter to a context manager, added jogging up/down to present the plate

# Review requests


# Risk assessment

When pickup/dropoff offsets get baked into shared-data, we need to be careful to include this 3.5mm offset.

fsinapi committed:
Remove gripper wiggle

fsinapi committed:
Add thermocycler driver functionality to move the lid an arbitrary number of degrees

fsinapi committed:
Pick up from "lifted" TC2 position:
- Change the lifting function to a context manager
- Move the lid up an extra 20 degrees before picking up, and then drop it back down after the grip action is over
- If the pickup location is a TC2, increase the Z offset by 3.5mm to account for the extra lift
- Adjusted tests to account for changes

fsinapi committed:
Added test for TC HC for raise/lower plate

fsinapi committed:
Merge branch 'edge' into RCORE-751-Pick-up-from-lifted-thermocycler-position

fsinapi committed:
Remove extra 3.5mm offset from pickup position - we will handle this in scripts instead

fsinapi committed:
Remove unnecessary state wait for the plate position because it will stay at "open"

fsinapi committed:
Remove unnecessary `is_tc2` check

fsinapi committed:
Remove unnecessary import

sanni-t commented on api/src/opentrons/protocol_engine/execution/thermocycler_plate_lifter.py:
```diff
@@ -48,11 +43,34 @@ async def lift_plate_for_labware_movement(
+        """Lift plate if moving labware from Thermocycler Gen2.
+
+        The 'lift_plate' function unsticks the plate, but does not leave the plate lifted.
+        In order to
```
incomplete sentence?

sanni-t commented on api/src/opentrons/protocol_engine/execution/labware_movement.py:
```diff
@@ -108,88 +108,68 @@ async def move_labware_with_gripper(
+                additional_offset_vector=experimental_offset_data.pickUpOffset,
+            )
 
-        if isinstance(current_location, ModuleLocation):
-            # TODO: This is a nasty hard-coded implementation to add a wiggle
-            # to dislodge plates from *all* modules. This should be changed to
-            # check configuration values on a per-module basis for 1) whether
-            # to wiggle at all 2) how much to wiggle.
-            # See https://opentrons.atlassian.net/browse/RCORE-749
-            WIGGLE_Z_LIFT = 0.5
-            WIGGLE_MAG = 0.5
-            # Lift the gripper a tiny bit in the Z axis
-            await ot3api.move_rel(mount=gripper_mount, delta=Point(z=WIGGLE_Z_LIFT))
```
This was doing the wiggle for all modules. From the fact that you removed it, am I understanding correctly that it was necessary _only_ for the thermocycler?

fsinapi replied:
There was a little bit of concern about picking up from a very cold TempDeck, but that never really materialized into a real gripper issue even without the wiggle so the thinking is that it probably isn't needed.

sanni-t replied:
And assuming it's not needed for mag block or heater-shaker either?
..Oh.. was the concern that the coldness would make a labware get stuck on the adapter so it'll need a wiggle to unstick?

fsinapi replied:
Yep, the condensation was the main concern and it doesn't seem to be an issue.

The TC mostly has sticking issues due to the intense force that the lid exerts on the plate, none of the other modules experience that kind of force so the plates don't tend to stick to t hem.

sanni-t commented on api/src/opentrons/hardware_control/modules/thermocycler.py:
```diff
@@ -233,6 +233,22 @@ async def lift_plate(self) -> str:
+        # remain at "open"
+
+    async def return_from_raise_plate(self) -> None:
+        """Move lid an extra bit."""
```
maybe
```suggestion
        """Return lid back to normal open position."""
```

sanni-t commented on api/tests/opentrons/protocol_engine/execution/test_labware_movement_handler.py:
```diff
@@ -171,6 +171,12 @@ async def test_move_labware_with_gripper(
         state_store.labware.get_labware_offset_vector("my-teleporting-labware")
     ).then_return(LabwareOffsetVector(x=0.1, y=0.2, z=0.3))
 
+    decoy.when(
+        thermocycler_plate_lifter.lift_plate_for_labware_movement(
+            labware_location=from_location
+        )
+    ).then_enter_with(value=None)
```
Hmm.. is this enough to verify that `lift_plate_for_labware_movement` context was actually entered? As in, since this mock is returning None (understandably), there is no dependency that will error out in case it's not entered. If the `move_labware` method doesn't enter the lift plate context, then does this test fail or does it just raise a warning that the mock was never entered? (asking since I have not used this context manager mocking functionality).

fsinapi replied:
Yeah looks like you're correct, this doesn't actually verify that the mock is entered!

What works better is for the mock to return another mock, and verify that `__aenter__` is called on that mock.

sanni-t replied:
Nice, that sounds much better!

sanni-t reviewed (approved):
This is an interesting technique! I think it also makes it harder (if not impossible) for the gripper to collide with the lid's pins so that's a nice side-effect.
This looks good to me mostly. Just some docstring typos and a question about context manager testing.

sfoster1 commented on api/src/opentrons/protocol_engine/execution/labware_movement.py:
```diff
@@ -108,88 +108,68 @@ async def move_labware_with_gripper(
-            await ot3api.move_rel(mount=gripper_mount, delta=Point(y=WIGGLE_MAG))
-            await ot3api.move_rel(mount=gripper_mount, delta=Point(y=-2 * WIGGLE_MAG))
-            await ot3api.move_rel(mount=gripper_mount, delta=Point(y=WIGGLE_MAG))
+            for waypoint in waypoints_to_labware:
```
nit: i sort of hate this internal comparison thing; maybe
```
for waypoint in waypoints_to_labware[:-1]:
    await ot3api.move_to(mount=gripper_mount, abs_position=waypoint)
# TODO: ...
await ot3api.home_gripper_jaw()
await ot3api.move_to(mount=gripper_mount, abs_position=waypoints_to_labware[-1])
``` 
? 
Does the same thing but looks nicer IMO

fsinapi replied:
Don't disagree, but also this code isn't being changed in this PR other than indenting it with the context manager.

sanni-t replied:
That whole file needs a refactor but ya any small improvements made everytime that code is touched would be appreciated.

fsinapi replied:
ok, [added this](https://github.com/Opentrons/opentrons/pull/12797/commits/ccb3eaa63ff6fc996815fed1cd70b2301be1d0ad)!

fsinapi committed:
Fixed incomplete doc sentence

sfoster1 commented on api/src/opentrons/hardware_control/modules/thermocycler.py:
```diff
@@ -233,6 +233,22 @@ async def lift_plate(self) -> str:
+        if not self.lid_status == ThermocyclerLidStatus.OPEN:
+            raise RuntimeError("Lid must be open")
+        await self.wait_for_is_running()
+        await self._driver.jog_lid(20)
```
I regret to announce that i'm on team "magic numbers should get a name" and I think these should be consts at the top somewhere

sfoster1 commented on api/src/opentrons/hardware_control/modules/thermocycler.py:
```diff
@@ -233,6 +233,22 @@ async def lift_plate(self) -> str:
+    async def return_from_raise_plate(self) -> None:
+        """Move lid an extra bit."""
+        await self.wait_for_is_running()
+        await self._driver.jog_lid(-23)
```
ditto here re: magic numbers

sfoster1 reviewed (changes requested):
I think we should give those jog amounts names and maybe make them absolute values just in case we have a state mismatch.

I'm also not super into the idea about putting the z offset in the protocols. If we're intending on later loading that offset into the protocol engine directly, then that will make the protocols incorrect. Meanwhile, we already have this `load_experimental_offset` function that is where we'd fold in that value from config anyway, so I really don't see the benefit in not putting the number there in practice.

fsinapi committed:
Move pre-TC home to only happen if this is a TC

fsinapi committed:
Cleaned up TC Plate Lift mock to actually test that the plate lift occurs

fsinapi committed:
Named constants

fsinapi commented:
> I'm also not super into the idea about putting the z offset in the protocols. If we're intending on later loading that offset into the protocol engine directly, then that will make the protocols incorrect. Meanwhile, we already have this `load_experimental_offset` function that is where we'd fold in that value from config anyway, so I really don't see the benefit in not putting the number there in practice.

AFAIK, _all_ of the offsets are in protocols right now, using the `pickUpOffset` and `dropOffset` parameters to the `MoveLabware` command (which ends up being in `experimental_offset_data`). So the thinking was that the `pickUpOffset` should get bumped by 3.5mm, and when [RLAB-295](https://opentrons.atlassian.net/browse/RLAB-295) is implemented then the pickup offset would live there.

fsinapi committed:
Update api/src/opentrons/hardware_control/modules/thermocycler.py

Co-authored-by: Sanniti Pimpley <sanni-t@users.noreply.github.com>

fsinapi committed:
Clean up iteration through waypoints

sanni-t commented:
> AFAIK, _all_ of the offsets are in protocols right now, using the `pickUpOffset` and `dropOffset` parameters to the `MoveLabware` command (which ends up being in `experimental_offset_data`). So the thinking was that the `pickUpOffset` should get bumped by 3.5mm, and when [RLAB-295](https://opentrons.atlassian.net/browse/RLAB-295) is implemented then the pickup offset would live there.

Yep, this is correct.
But, like @sfoster1 said, I wonder if it would actually be better to add this 3.5mm to the `pickupOffset` inside `get_experimental_labware_movement_offset_vector` at this point. If we push this responsibility to protocols, all science/testing teams will need to update their protocols. Then once we merge in RLAB-295 (which is slated for next week's release), everyone will have to update their protocols again. So even though it feels weird to have some offsets in protocol and some in engine, it might be an acceptable, non-breaking way to implement this. Or is there anything else in offset calculations I'm missing? 

If you do decide to add the 3.5 in engine itself, I'm fine hard-coding it in `get_experimental_labware_movement_offset_vector`. We'll take care of including it in the contextual offsets calculation when handling RLAB-295.

fsinapi committed:
Changed TC2 pickups to automatically add in the extra 3.5mm offset. Updated tests to check this as well.

fsinapi commented:
Alrighty @sanni-t @sfoster1 , updated the implementation to automatically bump the Z offset by 3.5mm when picking up from a TC2. Updated the PR description to match. Definitely feels like a less-annoying approach for anybody running protocols.

sfoster1 reviewed (approved):
I think this is good, nice! The offset stuff looks messy but it will look less messy once it's all in configuration.

fsinapi committed:
Fixed TC2 sim and related test for plate lifting
